sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/SingletonPatternCheck.kt

```diff
@@ -0,0 +1,131 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.descriptors.ConstructorDescriptor
+import org.jetbrains.kotlin.psi.KtBlockExpression
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtClassBody
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.psi.KtFunctionLiteral
+import org.jetbrains.kotlin.psi.KtLambdaArgument
+import org.jetbrains.kotlin.psi.KtLambdaExpression
+import org.jetbrains.kotlin.psi.KtModifierListOwner
+import org.jetbrains.kotlin.psi.KtObjectDeclaration
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.KtPropertyDelegate
+import org.jetbrains.kotlin.psi.KtReferenceExpression
+import org.jetbrains.kotlin.psi.allConstructors
+import org.jetbrains.kotlin.psi.psiUtil.visibilityModifier
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.AbstractCheck
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+import org.sonarsource.kotlin.visiting.KtTreeVisitor
+
+@Rule(key = "S6515")
+class SingletonPatternCheck : AbstractCheck() {
+
+    override fun visitKtFile(file: KtFile, kotlinFileContext: KotlinFileContext) {
+        val singletonClassCandidateExtractor = SingletonClassCandidateExtractor()
+        singletonClassCandidateExtractor.visitTree(file)
+        val singletonClassCandidates = singletonClassCandidateExtractor.singletonClassCandidates
+        if (singletonClassCandidates.isEmpty()) return
+
+        val singleConstructorCallExtractor =
+            SingleConstructorCallExtractor(singletonClassCandidates, kotlinFileContext.bindingContext)
+        singleConstructorCallExtractor.visitTree(file)
+        val singleConstructorCallByClass = singleConstructorCallExtractor.singleConstructorCallByClass
+        if (singleConstructorCallByClass.isEmpty()) return
+
+        singleConstructorCallByClass.values.stream().filter {
+            isInitializingCall(it) || isLazyInitializingCall(it, kotlinFileContext.bindingContext)
+        }.forEach {
+            kotlinFileContext.reportIssue(it, "Singleton pattern should use object declarations or expressions")
+        }
+    }
+}
+
+private class SingletonClassCandidateExtractor : KtTreeVisitor() {
+
+    val singletonClassCandidates = mutableSetOf<String>()
+
+    override fun visitClass(klass: KtClass) {
+        if (isSingletonClassCandidate(klass)) {
+            klass.fqName?.let {
+                singletonClassCandidates.add(it.toString())
+            }
+        }
+    }
+}
+
+private fun isSingletonClassCandidate(klass: KtClass): Boolean {
+    val constructors = klass.allConstructors
+    return klass.isPrivate() || (constructors.isNotEmpty() && constructors.all(KtModifierListOwner::isPrivate))
```

Comment:
From Java experience I have the feeling that singletons usually only declare one no-arg constructor (or none, although that case may be harder to handle, as we can't find if it is instantiated elsewhere). So maybe we can reduce our scope here a little and ignore classes with multiple constructors (why have them if one isn't called?). Also, why pass arguments to a constructor if they are always the same? This can also help us later to determine the number of invocations more efficiently. I am happy to discuss this point though, perhaps it is different in Kotlin (I am not sure).